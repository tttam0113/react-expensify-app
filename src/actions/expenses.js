import uuid from 'uuid';
import database from '../firebase/firebase';

export const addExpense = expense => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return dispatch => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;

        const expense = { description, note, amount, createdAt };

        return database
            .ref('expenses')
            .push(expense)
            .then(ref => {
                dispatch(addExpense({ id: ref.key, ...expense }));
            });
    };
};

export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//1.create startRemoveExpense 9same call signature as removeExpense)
//2.Test startRemoveExpense with "should remove expenses from firebase"
//3.Use startRemoveExpense in EditExpensePage instead of removeExpense
//4.Adjust EditExpensePage tests
export const startRemoveExpense = ({ id } = {}) => {
    return dispatch => {
        return database
            .ref(`expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(removeExpense({ id }));
            });
    };
};

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

//SET_EXPENSES
export const setExpenses = expenses => ({
    type: 'SET_EXPENSES',
    expenses
});

//1.fetch all expense data onse
//2. Parse that data into an array
//3. Dispatch SET_EXPENSES

export const startSetExpenses = () => {
    return dispatch => {
        return database
            .ref('expenses')
            .once('value')
            .then(snapshot => {
                const expenses = [];
                snapshot.forEach(childSnapshot => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });

                dispatch(setExpenses(expenses));
            });
    };
};
