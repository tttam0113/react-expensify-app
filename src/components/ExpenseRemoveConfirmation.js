import React from 'react';
import Modal from 'react-modal';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseRemoveConfirmation = ({
    isOpen,
    expense,
    onRemoveClick,
    onCancelClick
}) => (
    <Modal
        className="modal"
        isOpen={isOpen}
        onRequestClose={onCancelClick}
        contentLabel="Remove Expense"
        closeTimeoutMS={200}
        ariaHideApp={false}
    >
        <h2 className="modal__title">Are you sure want to remove this expense?</h2>
        <div className="modal__body">
            <div>
                <h3 className="list-item__title">{expense.description}</h3>
                <span className="list-item__sub-title">
                    {moment(expense.createdAt).format('MMMM Do, YYYY')}
                </span>
            </div>

            <h3 className="list-item__data">
                {numeral(expense.amount / 100).format('$0,0.00')}
            </h3>
        </div>
        <div className="modal__actions">
            <button className="button" onClick={onCancelClick}>No</button>
            <button className="button button--alert" onClick={onRemoveClick}>Yes</button>
        </div>
    </Modal>
);

export default ExpenseRemoveConfirmation;
