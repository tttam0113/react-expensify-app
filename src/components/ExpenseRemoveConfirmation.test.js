import React from 'react';
import { shallow } from 'enzyme';

import ExpenseRemoveConfirmation from './ExpenseRemoveConfirmation';
import expenses from '../tests/fixtures/expenses';

let onRemoveClick, onCancelClick, wrapper;

beforeEach(() => {
    onRemoveClick = jest.fn();
    onCancelClick = jest.fn();
    wrapper = shallow(
        <ExpenseRemoveConfirmation
            isOpen={true}
            onRemoveClick={onRemoveClick}
            onCancelClick={onCancelClick}
            expense={expenses[2]}
        />
    );
});

test('should render ExpenseRemoveConfirmation correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should call onCancelClick when click outsite modal', () => {
    wrapper.find('Modal').prop('onRequestClose')();
    expect(onCancelClick).toHaveBeenCalled();
});

test('should call onCancelClick when click no button', () => {
    wrapper.find('button').first().simulate('click');
    expect(onCancelClick).toHaveBeenCalled();
});

test('should call onRemoveClick when click yes button', () => {
    wrapper.find('.button--alert').simulate('click');
    expect(onRemoveClick).toHaveBeenCalled();
});
