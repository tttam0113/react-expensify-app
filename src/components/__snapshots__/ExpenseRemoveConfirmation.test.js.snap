// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseRemoveConfirmation correctly 1`] = `
<Modal
  ariaHideApp={false}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  contentLabel="Remove Expense"
  isOpen={true}
  onRequestClose={[MockFunction]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <h2
    className="modal__title"
  >
    Are you sure want to remove this expense?
  </h2>
  <div
    className="modal__body"
  >
    <div>
      <h3
        className="list-item__title"
      >
        Credit Card
      </h3>
      <span
        className="list-item__sub-title"
      >
        January 5th, 1970
      </span>
    </div>
    <h3
      className="list-item__data"
    >
      $45.00
    </h3>
  </div>
  <div
    className="modal__actions"
  >
    <button
      className="button"
      onClick={[MockFunction]}
    >
      No
    </button>
    <button
      className="button button--alert"
      onClick={[MockFunction]}
    >
      Yes
    </button>
  </div>
</Modal>
`;
